rollup-interface/src/state_machine/da.rs

```diff
@@ -56,14 +58,68 @@ pub trait DaVerifier {
     ) -> Result<(), Self::Error>;
 }
 
+// Simple structure that implements the Read trait for a buffer and  counts the number of bytes read from the beginning
+// Useful for the partial blob reading optimization: we know for each blob how many bytes have been read from the beginning
+// Because of soundness issues we cannot implement the Buf trait because the prover could get unproved blob data using the chunk method.
+#[derive(Debug, Clone, Serialize, Deserialize, BorshDeserialize, BorshSerialize, PartialEq)]
+pub struct BufReaderWithCounter<B: Buf> {
+    inner: B,
+    acc: Vec<u8>,
+    counter: usize,
+}
+
+impl<B: Buf> BufReaderWithCounter<B> {
+    pub fn new(inner: B) -> Self {
+        let buf_size = inner.remaining();
+        BufReaderWithCounter {
+            inner,
+            counter: 0,
+            acc: Vec::with_capacity(buf_size),
+        }
+    }
+
+    pub fn counter(&self) -> usize {
+        self.counter
+    }
+
+    pub fn acc(&self) -> &Vec<u8> {
+        &self.acc
+    }
+}
+
+impl<B: Buf> Read for BufReaderWithCounter<B> {
+    // Reads the inner buf into the provided buffer, and appends the data read to inner accumulator
+    fn read(&mut self, mut buf: &mut [u8]) -> std::io::Result<usize> {
+        let len_before_reading = self.inner.remaining();
+
+        buf.write_all(&self.inner.copy_to_bytes(min(buf.len(), len_before_reading)))?;
+
+        let num_read = len_before_reading - self.inner.remaining();
+
+        self.acc.append(&mut buf.to_vec());
```

Comment:
We're doing an extra copy here. Let's use `extend_from_slice` so that we don't have to convert the buf to a `vec`